src/internals/helpers/math.ts

```diff
@@ -0,0 +1,24 @@
+import { ValueError } from "@/errors.js";
+
+export function cosineSimilarity(vecA: number[], vecB: number[]): number {
+  if (vecA.length !== vecB.length) {
+    throw new ValueError("Vectors must have equal length");
+  }
+  // (cos θ) = (A · B) / (|A| * |B|)
+  const dot = vecA.reduce((sum, value, index) => sum + value * vecB[index], 0);
+  const magA = Math.sqrt(vecA.reduce((sum, value) => sum + value ** 2, 0));
+  const magB = Math.sqrt(vecB.reduce((sum, value) => sum + value ** 2, 0));
+
+  if (magA === 0 || magB === 0) {
+    throw new ValueError("Vectors cannot not have zero magnitude");
+  }
+  return dot / (magA * magB);
+}
+
+export function cosineSimilarityMatrix(matrixA: number[][], matrixB: number[][]): number[][] {
+  if ((matrixA[0]?.length ?? 0) !== (matrixA[0]?.length ?? 0)) {
+    throw new Error("Matrices must have the same number of columns.");
```

Comment:
`ValueError` again please

Reply:
Resolved in [529ca86](https://github.com/i-am-bee/bee-agent-framework/pull/243/commits/529ca86a06f0495176997e6d7545e9cfa8a502fc)